ZOOKEEPER-837: Eliminate cycle dependency between ClientCnxn and ZooKeeper

TisonKun opened the pull request:
1. Extract ZKWatchManager to single file
2. Move ZKWatchManager instance to ClientCnxn in order to eliminate cycle dependency
3. let `ZooKeeper` syncs a copy of default watcher, in order to reduce dependencies to `getWatchManager()`

maoling commented:
Have Checkstyle violations, Looking

TisonKun commented:
Thanks for your reminder @maoling! Updated.

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -155,13 +155,11 @@
-    private final ZooKeeper zooKeeper;
-
-    private final ClientWatchManager watcher;
+    private final ZKWatchManager watchManager;
```
this one should be ClientWatchManager

TisonKun replied:
Nope *for now*. `ClientWatchManager` provides the only interface `materialize` while we need to retrieve dataWatches and so on. It can be a separated issue we revisit `ClientWatchManager` and enrich interfaces it provides, as well as loose this type constrain to `ClientWatchManager`.

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java:
```diff
@@ -605,7 +321,7 @@ public DataWatchRegistration(Watcher watcher, String clientPath) {
         @Override
         protected Map<String, Set<Watcher>> getWatches(int rc) {
-            return watchManager.dataWatches;
+            return getWatchManager().dataWatches;
```
dataWatches -> getDataWatches() 
otherwise we are breaking the incapsulation you are introducing

TisonKun replied:
good point. Addressing...

TisonKun commented:
Thanks for your review @eolivelli! Addressing inline comments...

These changes could bring impact to forks if they want to sync with this repo. However, it should not break "downstream projects" IMO because where `ZKWatcherManager` places is internal for ZooKeeper and there is no interfaces break. Do you know any downstream project programs directly to this internal concept?

Besides, is there a plan towards ZK 4.x?

maoling commented on zookeeper-server/src/main/java/org/apache/zookeeper/ZKWatchManager.java:
```diff
@@ -0,0 +1,297 @@
+ * We are implementing this as a nested class of ZooKeeper so that
+ * the public methods will not be exposed as part of the ZooKeeper client
+ * API.
+ */
```
this copied annotate needs a update?

TisonKun replied:
Thanks for your review. I'm going to update it and rebase the pull request in this week.

maoling commented:
> it should not break "downstream projects" IMO because where ZKWatcherManager places is internal for ZooKeeper and there is no interfaces break.

Agreed with this. Uses cannot include `import org.apache.zookeeper.ZooKeeper.ZKWatchManager` explicitly (test for this).

TisonKun commented:
@maoling I've updated the pull request and rebased on current master

@hanm  @anmolnar I'm not sure whether or not it should be in zk 4.x since @eolivelli asked for that. Could you please take a look at this thread and share your ideas? If we have to make it in zk 4.x, I will postpone effort and revive it when the develop loop starts.

hanm commented:
This change seems OK with me for 3.x. It's API backward compatible. 

For binary compatibilities, it might not (not sure... have to look jdk docs) - but I don't think we strive for binary compatibilities anyway (e.g there were a few previous releases were not binary compatible with older version due to new apis added, which requires client code to recompile).

TisonKun commented:
Thanks for your insight @hanm !

TisonKun commented:
JenkinsMaven fails as https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/1624/

However, the page doesn't show any info about the failure, even it reports no failures. I have no idea what happens and what I can do.

eolivelli commented:
error: openssl/ossl_typ.h: No such file or directory
     [exec]  #include <openssl

We are missing openssl on jenkins

TisonKun commented:
@eolivelli I don't think I can do something for it 😷

eolivelli commented:
Just wait, we are going to fix it

TisonKun commented:
Thanks for your info and works @eolivelli . I'll appreciate it if you ping me on fixed.

TisonKun commented:
retest this please

TisonKun commented:
@maoling @hanm @eolivelli @anmolnar I think this pull request is mergeable. Please take a look.

maoling commented on zookeeper-server/src/main/java/org/apache/zookeeper/ZKWatchManager.java:
```diff
@@ -0,0 +1,443 @@
+ * as part of ZooKeeper client API.
+ */
+class ZKWatchManager implements ClientWatchManager {
+
```
Great

maoling commented on zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java:
```diff
@@ -2268,7 +1885,7 @@ public Stat exists(final String path, Watcher watcher) throws KeeperException, I
     public Stat exists(String path, boolean watch) throws KeeperException, InterruptedException {
-        return exists(path, watch ? watchManager.defaultWatcher : null);
+        return exists(path, watch ? defaultWatcher : null);
     }
```
> let ZooKeeper syncs a copy of default watcher, in order to reduce dependencies to getWatchManager()

 You have used `getWatchManager()` in the `zookeeper.java`, `getWatchManager().getDefaultWatcher()` is better? The origin logic is: get the `defaultWatcher` from `ZKWatchManager`.

TisonKun replied:
Yes definitely. Updating...

maoling commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/NettyServerCnxnTest.java:
```diff
@@ -85,10 +85,9 @@ public void tearDown() throws Exception {
-        final String path = "/a";
-        try {
+        try (ZooKeeper zk = createClient()) {
+            final ZooKeeperServer zkServer = serverFactory.getZooKeeperServer();
```
Please don't do unrelated refactor work next time

TisonKun replied:
Sorry for that. I will keep in mind.

TisonKun commented:
Any news here? Still I would say if it is undesirable to got in 3.x I will freeze this thread.

@hanm @eolivelli

TisonKun commented:
@eolivelli Thanks for you information. I'm glad to release the resource of our community from reviewing this now. However, I'm curious how it break clients? Since others here have the idea that it is API compatible, is there a concrete example to show the problem?

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -352,35 +354,29 @@ public String toString() {
         int sessionTimeout,
-        ZooKeeper zooKeeper,
-        ClientWatchManager watcher,
+        ZKClientConfig clientConfig,
```
this is my major concern.

It is true that we are not changing the very high level API, as ClientCnxn is an internal class and applications should not use it directly.

But it is probable that frameworks like Curator (ping @Randgalt ) do tricks over this stuff.

This change is a big refactor and touches the very internals of the client so we should handle it with care.

Don't get my wrong, I think this work is awesome and I hope we can make it land it to master someday.

That's only my opinion, let's see if any other reviewer wants to sponsor this change

Randgalt replied:
On first glance it doesn't look like it conflicts with Curator. But, maybe @TisonKun can test with Curator locally. As an aside, I wonder if we shouldn't include some Curator test in ZooKeeper in the future - i.e. validate that new changes don't break Curator.

TisonKun replied:
@eolivelli After a close look I find that where Curator hacking into `CliengCnxn` is only `InjectSessionExpiration` that get access of `cnxn`/`eventThread`/`queueEvent`. So this patch should not break Curator.

And I second @Randgalt that it is Curator responsibility to adjust its trick for accessing ZooKeeper internal concept since they don't promise compatibility. IMO Curator already has a `Compatibility` utils for different manner between zk34 and zk35. We should continue refining codebase so that projects such as Curator don't have to rely on internal concepts, though.

eolivelli commented:
We can merge this patch for 3.7.0

Thank you all for double checking

eolivelli commented:
@TisonKun can you please rebase ?
We should be ready for this patch now

we also need another binding +1, but having CI green and patch ready will be a good preparation for it

TisonKun commented:
Thanks for picking it up @eolivelli ! I will schedule rebasing this patch in days.

TisonKun commented:
@eolivelli rebased.

eolivelli commented:
It looks like Travis failed

TisonKun commented:
@eolivelli I've resolve license issue but it seems CI stuck.

TisonKun commented on zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -1010,11 +1002,11 @@ void primeConnection() throws IOException {
             // TODO: here we have the only remaining use of zooKeeper in
             // this class. It's to be eliminated!
```
Exactly! I think the removal lost in the last rebase :/

Updated.

TisonKun commented:
Hi @eolivelli it seems we almost reach a green status to include this patch. PTAL :-)

eolivelli commented:
Yes I will merge as soon as possible

TisonKun commented:
ping @eolivelli as a reminder :-)

eolivelli commented:
Done! 
Thanks for the reminder

TisonKun commented:
Thanks!
